### PR TITLE
[Merged by Bors] - Resources limits

### LINKED ISSUE #128: Implement resource limits for Spark pods
In other products we're introducing a common resource limit configuration. This ticket is about evaluating whether it makes sense to use it in this operator too. If it does, this ticket includes the implementation.

### PR DESCRIPTION
**Pull request by adwk67**
# Description

Define resource limits for job, driver and executor pods.
Fixes #128 

## Review Checklist

- [x] Code contains useful comments
- [x] CRD change approved (or not applicable)
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/crd/src/lib.rs [hunk 1]
```diff
@@ -48,6 +53,41 @@ pub struct SparkApplicationStatus {
...
+}
+
+#[derive(Clone, Debug, Default, Deserialize, JsonSchema, PartialEq, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct SparkConfig {
+    pub resources: Option<Resources<SparkStorageConfig, NoRuntimeLimits>>,
+}
+
+impl SparkConfig {
+    fn default_resources() -> Resources<SparkStorageConfig, NoRuntimeLimits> {
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:70]: I suggest making the default resources depending on the role (job, driver, executor)
The job needs basically nothing (maybe 50m/100m cou, 1gb ram), default for driver i would suggest something like 100m/1 CPU and 2gb ram and executors 1/4 cpu and 4gb ram
Just throwing in some numbers to give an idea, no need to stick to them exactly

Comment (adwk67) [reply rust/crd/src/lib.rs:70]: Good idea! Yes, I will do that

Comment (adwk67) [reply rust/crd/src/lib.rs:70]: Done.

File: tests/templates/kuttl/resources/13-assert.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+apiVersion: kuttl.dev/v1beta1
+kind: TestAssert
+timeout: 240
+skipLogOutput: true
+commands:
+  - script: kubectl get cm -n $NAMESPACE spark-resources-pod-template -o yaml | yq -r '.data."driver.yml"' | yq -r '.spec.containers[0].resources.limits.cpu' | grep 1500m
+  - script: kubectl get cm -n $NAMESPACE spark-resources-pod-template -o yaml | yq -r '.data."driver.yml"' | yq -r '.spec.containers[0].resources.requests.cpu' | grep 1
+  - script: kubectl get cm -n $NAMESPACE spark-resources-pod-template -o yaml | yq -r '.data."executor.yml"' | yq -r '.spec.containers[0].resources.limits.cpu' | grep 2
+  - script: kubectl get cm -n $NAMESPACE spark-resources-pod-template -o yaml | yq -r '.data."executor.yml"' | yq -r '.spec.containers[0].resources.requests.cpu' | grep 1
```
Comment (razvan) [inline tests/templates/kuttl/resources/13-assert.yaml:10]: Can this not be done declaratively ? Something like:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: spark-resources-pod-template
spec:
  containers:
    - resources:
        - requests:
            cpu: 1
...
```

**Review by razvan — COMMENTED** [review]
Tests pass and it looks good. One question though.

File: tests/templates/kuttl/resources/13-assert.yaml (hunk 2 above)
Comment (adwk67) [reply tests/templates/kuttl/resources/13-assert.yaml:10]: hmmmm, das scheint leider nicht zu funktionieren, da das ganze Data-Element wird verglichen, inklusiv von UIDs:

```
         ...
        != actual: ---
        apiVersion: v1
        kind: Pod
        metadata:
          labels:
            app.kubernetes.io/instance: spark-resources
            app.kubernetes.io/managed-by: spark-k8s-operator
            app.kubernetes.io/name: spark-k8s
            app.kubernetes.io/version: "1.0"
          name: spark-driver
          ownerReferences:
            - apiVersion: spark.stackable.tech/v1alpha1
              kind: SparkApplication
              name: spark-resources
              uid: 6841c21a-50a2-4969-b529-665702180c26
        ...
```

**Review by razvan — APPROVED** [review]

**Comment by adwk67** [comment]
bors merge

**Pull request closed without merging**